[tests] Add support for running tests on helix

radical opened the pull request:
## Why?

Using helix to run tests would mainly be useful for End2End integration tests which would need controlled versions of dependencies, and that would be easier to do on helix than the build machines.  This could be used to setup custom docker containers, and networks to simulate various scenarios and run e2e tests.

## With this PR, tests will be run in two steps on CI:

1. On the build machine - which is the same as they are run currently.
    This will be run in a single step, from the source repo with built
 `artifacts`.

2. On helix - for this the tests are archived to `.zip` files, and then
     sent to helix with any dependencies.

- all the tests will still continue to use `dotnet-coverage`. But since
tests can run on more than one machine now, there is an additional CI
step that merges all the coverage xml files.

### Details

This adds new properties that affect CI builds:

- `$(IsTestSupportProject)` - which is for a project that should be run
  as a test project itself.
- `$(RunTestsOnHelix)` - which defaults to `false`

- Also, a default `.runsettings` file is used to allow easier setting of
  loggers, and other configuration for the tests.
- The dashboard tests are explicitly marked for running on helix, just so we have at least one such test suite. This PR explicitly avoids supporting zero helix tests to ensure that the support does not rot, and helix step cannot silently fail because nothing got zipped.

- Increase the timeout for CI from 30mins to 45mins.

radical committed:
[tests] Add support for running tests on helix

With this PR, tests will be run in two steps on CI:

1. On the build machine - which is the same as they are run currently.
   This will be run in a single step, from the source repo with built
   `artifacts`.

2. On helix - for this the tests are archived to a `.zip` file, and then
   sent to helix with any dependencies.

Additionally, tests can be marked to always be run on helix or the build
machine. Other tests can be run on either.

- a new property `$(DefaultToRunningTestsOnHelix)` when true would run
  all the tests that are not build-machine-only to run on helix.
  - when the property is !true, all the tests that can run build-machine
    will run there. This is the default.

This adds new properties that affect CI builds:

- `$(IsTestSupportProject)` - which is for a project that should be run
  as a test project itself.
- `$(IsHelixOnlyTestProject)` - such a test can only be run on helix.
- `$(IsNonHelixOnlyTestProject)` - such a project must always run on the
  build machine, IOW, with the source repo.

- Also, a default `.runsettings` file is used to allow easier setting of
  loggers, and other configuration for the tests.

radical committed:
[tests] Add the CI side of support for running tests on helix

All the tests will still continue to use dotnet-coverage. But since
tests can run on more than one machine now, there is an additional CI
step that merges all the coverage xml files.

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests

radical committed:
send-to-helix.proj: extract dotnet-coverage version from dotnet-tools.json

radical commented:
Ready for review

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests

radical committed:
CI: Increase timeout from 30->45mins, to account for helix waiting on agents

radical commented:
Helix step can be slow to get agents sometimes, so increased the timeout to 45mins.

joperezr commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -17,21 +17,57 @@ parameters:
+    - template: /eng/pipelines/templates/send-to-helix.yml
+      parameters:
+        HelixProjectPath: '$(Build.SourcesDirectory)/tests/send-to-helix.proj'
+        HelixProjectArguments: /p:RunWithCodeCoverage=true /p:RepoTestResultsPath=${{ parameters.repoTestResultsPath }}
```
Will these not clash given we are using the same directory for non helix tests and helix tests?

radical replied:
In case of helix they will only have the coverage xml files.

joperezr commented on eng/pipelines/templates/send-to-helix.yml:
```diff
@@ -0,0 +1,94 @@
+# Please remember to update the documentation if you make changes to these parameters!
```
Loved the docs :).

Did we get these from some other repo? Mainly asking to see if we need to try to keep in sync.

radical replied:
The template is from `eng/common/templates/steps/send-to-helix.yml`, slightly modified here.

joperezr replied:
mmm I see, is there a way to instead use that template file and pass in some parameters to accommodate for our small modifications? If possible, I'd love to reduce the amount of yaml that we have to maintain.

radical replied:
This `eng/common/templates/steps/send-to-helix.yml` is a sample template from arcade, and it's not used directly. But I can make changes in that file itself.

joperezr replied:
Not sure what you mean when you say a "sample template". Tipically the templates under eng\common\templates are there so that repos can use them if they want/need as part of their build process. I don't want to block this PR on this, but I do want to make sure this is fixed. Can we at least log a tracking bug so that we go and add the necessary hooks in dotnet/arcade repo so that we can later ingest those changes here, and remove this file entirely so that we instead leverage the files from eng\common?

radical replied:
Opened https://github.com/dotnet/arcade/issues/14470 .

joperezr commented on tests/Directory.Build.props:
```diff
@@ -10,6 +10,16 @@
     <NoWarn>$(NoWarn);1573;1591;1712</NoWarn>
+
+    <IsHelixOnlyTestProject Condition="'$(IsHelixOnlyTestProject)' == ''">false</IsHelixOnlyTestProject>
+    <IsNonHelixOnlyTestProject Condition="'$(IsNonHelixOnlyTestProject)' == ''">false</IsNonHelixOnlyTestProject>
```
Given the above is a boolean, do we actually need this property? Can't we just infer based on the above?

radical replied:
We have 3 possible values - 1. helix only; 2. build only; 3. any . The two properties represent (1), and (2), and both being false means that (3) - no restrictions on where the tests can run.

We could use a string property with some fixed values like `['helix', 'non-helix', 'any']`, but then you have the value strings in multiple places where you want to check, and that could get misspelled and cause tests to get silently skipped.

I'm not attached to any specific way of doing it though.

The name isn't great either, for the second property - `IsNonHelixOnlyTestProject`. An alternative would be `IsBuildMachineOnlyTestProject` sounds less clear.

eerhardt replied:
In dotnet/aspnetcore, we have C# attributes to control whether a test runs on Helix or not. 

https://github.com/dotnet/aspnetcore/blob/515987ac92f801fb995d88a20282a61571523aab/src/Testing/src/xunit/SkipOnCIAttribute.cs
https://github.com/dotnet/aspnetcore/blob/515987ac92f801fb995d88a20282a61571523aab/src/Testing/src/xunit/SkipOnHelixAttribute.cs
https://github.com/dotnet/aspnetcore/blob/515987ac92f801fb995d88a20282a61571523aab/src/Testing/src/xunit/SkipNonHelix.cs

Would it be possible to do this that way? Then we wouldn't need to worry about misspelling, since it would be C# code.

joperezr replied:
I see, yeah I think that as @eerhardt pointed out in a separate comment, I guess that the name is a bit confusing. I like his suggestion of dropping Only, or alternatively we can have something like `OnlyRunOnHelix` and `DoNotRunOnHelix` or something along those lines. Don't feel strongly about my suggestion haha, but I do think that the current version might be misleading.

joperezr commented on tests/Directory.Build.targets:
```diff
@@ -0,0 +1,16 @@
+    <Error Condition="'$(TestArchiveTestsDir)' == ''" Text="TestArchiveTestsDir property to archive the test folder must be set." />
+
+    <MakeDir Directories="$(TestArchiveTestsDir)" />
+    <ZipDirectory SourceDirectory="$(OutDir)"
```
Interesting, I would have imagined that zipping up build output would be standard and provided by the Helix SDK.

radical replied:
Helix allows you to point to a directory which it can then upload as a zip file. But doing it this way gives us a way to also indicate the tests meant for helix. This is the same technique followed in `dotnet/runtime`, though we don't have to do the same thing.

joperezr commented on tests/send-to-helix.proj:
```diff
@@ -0,0 +1,101 @@
+  <PropertyGroup>
+    <Language>msbuild</Language>
+
+    <_workItemTimeout>00:20:00</_workItemTimeout>
```
Just curious, does this include machine availability, or only execution?

radical replied:
Only execution on helix. The machine availability would be defined in the yml.

joperezr commented on tests/send-to-helix.proj:
```diff
@@ -0,0 +1,101 @@
+    <PropertyGroup>
+      <_DotNetToolJsonPath>$(RepoRoot).config/dotnet-tools.json</_DotNetToolJsonPath>
+      <_DotNetToolJsonContent>$([System.IO.File]::ReadAllText($(_DotNetToolJsonPath)))</_DotNetToolJsonContent>
+      <_DotNetCoverageVersionRegex>(?:"dotnet-coverage": *{.*?"version": *")([^"]*?)(?:")</_DotNetCoverageVersionRegex>
```
NIT: Not really important, but I think you can get rid of the first and third non capturing groups (meaning removing the parens and questionmark from (?:.......), and in theory you can also change the space matching from ` *` to `\s*` which would also match a new line. Didn't check but that might also mean you don't necessarily have to use Singleline. Anyway, not really important

radical replied:
re:groups, that makes sense. I had that before I figured out how to get a group value out! I'll make the changes.

joperezr commented on tests/Aspire.Components.Common.Tests/Aspire.Components.Common.Tests.csproj:
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsTestSupportProject>true</IsTestSupportProject>
```
NIT: in the tests/testproject seems like you defined this in Directory.Build.props (probably due to some race condition of wantint to set before parent's D.B.p). Don't you need the same here?

radical replied:
Yep, it should be in DB.props . The reason for that is:
https://github.com/radical/aspire/blob/af66351ffdc187cb707c242be1f30953d9e2d0d9/tests/Directory.Build.props#L18-L20

```xml
    <RunTestOnHelix Condition="'$(IsTestSupportProject)' != 'true' and
                                  ('$(IsHelixOnlyTestProject)' == 'true' or
                                    ('$(IsNonHelixOnlyTestProject)' != 'true' and '$(DefaultToRunningTestsOnHelix)' == 'true'))">true</RunTestOnHelix>
```

joperezr commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -17,21 +17,57 @@ parameters:
+        HelixProjectArguments: /p:RunWithCodeCoverage=true /p:RepoTestResultsPath=${{ parameters.repoTestResultsPath }}
+
+        ${{ if eq(parameters.isWindows, 'true') }}:
+          HelixTargetQueues: Windows.11.Amd64.Client.Open
```
I suppose these were the ones suggested by the dnc eng team?

radical replied:
They said that we should use whatever queues we need. These are the ones that `dotnet/runtime` uses.

radical committed:
Address review feedback from @ joperezr

radical commented:
I want to double check something about the test results, so please don't merge yet.

radical commented:
> I want to double check something about the test results, so please don't merge yet.

All good.

eerhardt commented on tests/Aspire.Hosting.Tests/Directory.Build.props:
```diff
@@ -1,4 +1,7 @@
 <Project>
+  <PropertyGroup>
+    <IsNonHelixOnlyTestProject>true</IsNonHelixOnlyTestProject>
```
The name feels weird. It would seem that `IsHelixOnlyTestProject=false` should be the same thing as this.

If we want to say "this test should never run on Helix", I would drop "Only" in the name. `IsNonHelixTestProject=true`.

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests-test

radical committed:
Address review feedback

.. and use a single property named `$(RunTestsOnHelix)` that defaults to
`false`.

radical committed:
Address review feedback, and modify the original template for send-to-helix.yml

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests-test

eerhardt commented on eng/pipelines/azure-pipelines.yml:
```diff
@@ -98,7 +98,7 @@ stages:
 
       - job: windows
-        timeoutInMinutes: 30
+        timeoutInMinutes: 45
```
😞

joperezr replied:
FWIW, 30 mins was already pretty low and super close to the time it actually takes to build the repo in AzDO, so we were going to have to do this sooner rather than later.

eerhardt replied:
![image](https://github.com/dotnet/aspire/assets/8291187/366bbf8c-92ac-413e-aee2-922ad5b551ed)

that's from a recent CI run.

eerhardt commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -22,16 +22,48 @@ steps:
+        HelixBuild: $(Build.BuildNumber)
+        HelixAccessToken: $(HelixApiAccessToken)
+
+    - ${{ if eq(parameters.isWindows, 'true') }}:
+      - pwsh: |
+            $_cov_filelist=$((Get-ChildItem -Path ${{ parameters.repoTestResultsPath }} -Recurse -Filter *.cobertura.xml -File | % { $_.FullName }) -join ' ')
+            echo "${{ parameters.dotnetScript }} dotnet-coverage merge $_cov_filelist -f cobertura -o ${{ parameters.repoTestResultsPath }}/$(Agent.Os)_$(Agent.JobName).cobertura.xml" > merge-cov.cmd
+            cmd /c merge-cov.cmd
+            rm merge-cov.cmd
+        displayName: Merge code coverage results
+
+    - ${{ if ne(parameters.isWindows, 'true') }}:
```
Why do we need to split here? Why can't the same script run on both windows and non-windows?

Might need a comment explaining why we are doing this.

radical replied:
I ran into my lack of pwsh skills :D

radical replied:
Looking at the yml that generates the report, it handles merging already. So, this can be simplified.

eerhardt commented on tests/send-to-helix.proj:
```diff
@@ -0,0 +1,101 @@
+    <DotNetCliPackageType>sdk</DotNetCliPackageType>
+
+    <GlobalJsonContent>$([System.IO.File]::ReadAllText('$(RepoRoot)global.json'))</GlobalJsonContent>
+    <DotNetCliVersion>$([System.Text.RegularExpressions.Regex]::Match($(GlobalJsonContent), '(%3F&lt;="dotnet": ").*(%3F=")'))</DotNetCliVersion>
```
Who reads this value?

radical replied:
This is ready by helix targets in Arcade.

eerhardt committed:
Minor PR feedback

eerhardt commented on tests/send-to-helix.proj:
```diff
@@ -0,0 +1,101 @@
+  <PropertyGroup>
+    <Language>msbuild</Language>
+
+    <_workItemTimeout>00:20:00</_workItemTimeout>
```
What's the difference between this and the `TestSessionTimeout` in the `.runsettings` file? Why are they 2 different values?

radical replied:
The timeout defined in the yml is for the CI *step*, so the whole step that sends items to helix and waits on them. It would include the time taken while waiting for agents to be assigned, and the time take to run the tests on helix.

The timeout here is only for the time taken on helix, per helix item.

The timeout for the step is larger mainly to account for helix wait times, which can vary depending on how busy the queue is.

eerhardt replied:
What is the timeout in the `.runsettings` file for?

radical replied:
> What is the timeout in the `.runsettings` file for?

IIUC, that is a timeout for individual tests.

Also, it should have been 5minutes but I had bumped it up when I was getting timeouts and forgot to revert that.

radical commented on eng/common/templates/steps/send-to-helix.yml:
```diff
@@ -5,6 +5,8 @@ parameters:
   HelixBuild: $(Build.BuildNumber)       # required -- the build number Helix will use to identify this -- automatically set to the AzDO build number
   HelixTargetQueues: ''                  # required -- semicolon-delimited list of Helix queues to test on; see https://helix.dot.net/ for a list of queues
   HelixAccessToken: ''                   # required -- access token to make Helix API requests; should be provided by the appropriate variable group
+  HelixProjectPath: ''                   # required -- path to the project file to build
```
Does this get overwritten by arcade updates? If so, then I'll move this to a separate file as before, and just reference this original template in a comment. Even if not that maybe not touching this file would be good.

radical replied:
@joperezr I'm reverting to the earlier way of using a copy of that file. Lemme know if you have an objections to that.

radical committed:
Instead of modifying the template from arcade, use a copy in pipelines/templates

radical committed:
send-to-helix.yml - minor cleanup

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests

radical committed:
Fix merging code coverage results

The second stage processes the *cobertura.xml files to generate a
report. And for that it collects these xml files from all the jobs.
Earlier commits in this PR did not have unique names across jobs, like
for linux, and windows. Which meant that the report generation step only
handled one set of the xml files.

Fix that by using the original unique filenames for the merged xml
generated from the helix, and non-helix tests.

eerhardt commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -22,16 +22,45 @@ steps:
               --settings $(Build.SourcesDirectory)/eng/CodeCoverage.config
+              --output ${{ parameters.repoTestResultsPath }}/NonHelixTests.cobertura.xml
+              "${{ parameters.buildScript }} -testnobuild -test -configuration ${{ parameters.buildConfig }} /bl:${{ parameters.repoLogPath }}/tests.binlog $(_OfficialBuildIdArgs)"
+      displayName: Run non-helix tests
```
Is it possible to run the non-helix tests and the helix tests in parallel? There shouldn't be a reason to serialize them, and that could speed up our CI legs.

radical replied:
We could add a wrapper project that builds both in parallel from msbuild. Should that be in a follow up PR?

eerhardt replied:
Why would we need to build them in parallel? I'm talking about running them in parallel.

Why would we need a wrapper project? Does azdo allow for running 2 pipeline tasks at the same time?

radical replied:
By "build" in parallel I meant run the Test, and send-to-helix targets in parallel.

radical replied:
> Does azdo allow for running 2 pipeline tasks at the same time?

AFAIK, this isn't supported.

joperezr replied:
Yeah, the thing that is allowed is to run two jobs or stages at the same time, but not steps within a job AFAIK. While I do think that could speed up things a bit, I also think it might be overkill to gain 9 mins or so that it takes in average to run the non helix tests first. I propose keeping things as is, and then perhaps look in the future here to see if this is something we need to invest on.

radical commented:
@akoeplinger would you mind taking a look at this PR to check that we are doing this in an idiomatic way?

radical commented:
Should we go ahead with this?

joperezr reviewed (approved):
Changes LGTM. Thanks @radical!

joperezr reviewed (approved):
Changes LGTM. Thanks @radical!

radical committed:
Merge remote-tracking branch 'origin/main' into helix-tests

radical commented:
I have merged `main` into the branch, and will wait for the CI results to confirm that this didn't break since the last build 2 days ago.
